src/gemdat/transitions.py

```diff
@@ -401,3 +402,210 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+    percolate_z : bool
+        If True, consider paths that percolate along the z dimension
+
+    Returns
+    -------
+    total_energy_cost: float
+        Total energy cost of the best percolating path
+    best_starting_point: tuple
+        Coordinates of the starting site where the best percolating path was found
+    best_perc_path: list
+        List of coordinates of the path
+    best_perc_path_energy: list
+        Energy along the path
+    """
+    X_real, Y_real, Z_real = F.shape
+
+    # Find percolation using virtual images along the required dimensions
+    if not (percolate_x + percolate_y + percolate_z):
+        print('Warning: percolation is not defined')
+        return None, None, None
+    if percolate_x:
+        extended_F = np.empty((2 * F.shape[0], F.shape[1], F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:X_real, :, :] = F
+        extended_F[X_real:, :, :] = F
+        F = extended_F
+    if percolate_y:
+        extended_F = np.empty((F.shape[0], 2 * F.shape[1], F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:, :Y_real, :] = F
+        extended_F[:, Y_real:, :] = F
+        F = extended_F
+    if percolate_z:
+        extended_F = np.empty((F.shape[0], F.shape[1], 2 * F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:, :, :Z_real] = F
+        extended_F[:, :, Z_real:] = F
+        F = extended_F
+    X, Y, Z = F.shape
+
+    # Get F on a graph
+    F_graph = free_energy_graph(F, max_energy_threshold=1e7, diagonal=True)
+
+    # Find the lowest cost path that percolates along the x dimension
+    total_energy_cost = float('inf')
+    best_starting_point = None
+    best_perc_path = []
+
+    # Iterate only over the peaks that have not been part of a percolating path
+    peaks_mask = np.zeros(peaks.shape[0], dtype=bool)
+    remaining_peaks = len(peaks)
+
+    for counter, starting_point in enumerate(peaks):
+        if peaks_mask[counter]:
+            continue
+
+        # Get the end point which is a periodic image of the peak
+        end_point = (starting_point[0] + X_real * percolate_x,
+                     starting_point[1] + Y_real * percolate_y,
+                     starting_point[2] + Z_real * percolate_z)
+
+        # Find the shortest percolating path through this peak
+        try:
+            path, path_energy = optimal_path(
+                F_graph,
+                tuple(starting_point),
+                tuple(end_point),
+            )
+        except nx.NetworkXNoPath:
+            continue
+
+        # Calculate the path cost
+        path_cost = np.sum(path_energy)
+
+        # Remove all the peaks that have been part of the path
+        peaks_mask[counter] = True
+        for mask in range(counter + 1, len(peaks)):
+            if not peaks_mask[mask] and tuple(peaks[mask]) in path:
+                peaks_mask[mask] = True
+                remaining_peaks -= 1
```

Comment:
Is this correct? Can a peak not be part of a different path with a different starting point and have a total lower energy?

Reply:
You are right! I have removed this mask.